### ISSUE TITLE
How to disable RegExp.$1, etc. for proper subclasses of RegExp

### ISSUE THREAD

**Issue by claudepache** [claudepache/es-regexp-legacy-static-properties#6]
For concreteness, consider:

``` js
class MyRegExp extends RegExp {  }
let rx = new MyRegExp('(foo)');
rx.exec('foobar'); // <-- what action should be taken w.r.t. RegExp.$1, etc.
RegExp.$1; // <-- what should evaluate this expression to?
```

If we want to disable `RegExp.$1` here, there are three options:
1. Do nothing. — `RegExp.$1` will evaluate to a value captured during the preceding successful match done with a plain regexp.
2. Empty `RegExp.$1`, etc. — `RegExp.$1` will evaluate to the empty string.
3. Invalidate `RegExp.$1`, etc. — Evaluating `RegExp.$1` will throw a TypeError.

Comments:
- Option 1 is the simplest, but the worst for the programmer, as RegExp.$1 will lead to some randomly false result.
- Option 3 is better for debugging purposes, but might be dangerous for backward-compatibility, as buggy errant references to RegExp.$1 may start to randomly throw.
- Option 2 is a trade-off: it avoids the hazard of Option 3, and it provides false but consistent results for use with proper subclasses of RegExp.

Although Option 3 may be idealistically better, it is probably more prudent to use Option 2.

---

Moreover, when evaluating  `MyRegExp.$1`, two options seem reasonable:
<ol start=3><li>Always throw a TypeError
<li>Always return undefined.
</ul>

**Comment by claudepache** [comment]
Currently, dumb Option 1 is specced. However, `MyRegExp.$1` will now invariably return undefined.

**Comment by erights** [comment]
I don't understand. The two parts of your comment seem to contradict each other:

> Currently, dumb Option 1 is specced

vs

> However, MyRegExp.$1 will now invariably return undefined.

But this does suggest an option 4:

Have MyRegExp.$1 always return `undefined`, in order to be as compatible as possible with the absence of these static properties. The most common feature testing pattern would then perceive these features as being absent. Unfortunately, less common but more correct feature testing patterns, i.e., `"$1" in MyRegExp` would falsely perceive the feature as present. An even less common but still correct own check would again perceive the feature as absent.

This suggests an option 5: To make feature testing reliably indicate absence, we could make RegExp subclasses be exotic objects that report an absence of these properties even though these properties would non-exotically be inherited from their `[[Prototype]]`. Before considering this, we'd need to be sure that this illusion does not violate the object invariants, and that it could be emulated by a proxy.

Option 5 seems much too complex given the minor benefit. Altogether, I favor option 3. There are two levels of legacy compatibility concerns we must always keep in mind:
- **don't break the web** only involves whole pages/frames/realms. None of these have RegExp subclasses, and so are compat with all of these options, including option 3.
- **invariants assumed by maintained code that is linked with other code**. For these, any invariants promised in one version of the spec should generally be upheld by later versions of the spec. Since RegExp statics have never been in the spec, option 3 would not be a problem. On the other hand, this does argue against option 5.

**Comment by claudepache** [comment]
> I don't understand. The two parts of your comment seem to contradict each other

Sorry for the confusion. In the original scenario of this issue, a match is done using a `MyRegExp` instance, but information about the match is attempted to be retrieved from a static property of the `RegExp` constructor. For _that_ scenario, Option 1 is currently what is observed. In case the coder  attempts to read the property of the `MyRegExp` constructor instead, your Option 4 is indeed better, and is currently specced.

**Comment by erights** [comment]
Thanks. I am ok with option 4 but still prefer option 3.

**Comment by claudepache** [comment]
> but still prefer option 3.

So do I, but I wanted the advice of other people, but there doesn't seem to be much interest...

One invariant of `RegExp.$1`, is that it always returns a string. With Option 3, that invariant is broken (i.e., sometimes it throws an error). That should not affect most programs, because `RegExp.$1` is normally evaluated only after a successful match. However, there is still a small chance of breakage when mixing new code with legacy code: some previously benign bug could become fatal.

**Comment by erights** [comment]
Such mixing does not happen on old legacy pages because of the lack of RegExp subclasses. For new code, including new linking with old libraries, throwing an exception is the friendlier development signal. Any approach that does not throw has the greater (and more fatal) hazard of having development proceed under the silent violation of assumptions that seem to be true elsewhere. This is especially dangerous if we return an empty string rather than `undefined`.

If we both prefer option 3 and others are not showing much interest, let's specify option 3. If others then show interest, we can reevaluate in light of the interest they express.

**Comment by claudepache** [comment]
Fixed by 0a95955b5e2f13cd8053b0958cae505c4c19a6c9
